Update elastic-agent-providers.asciidoc (#520)
elastic/observability-docs: docs/en/ingest-management/elastic-agent/elastic-agent-providers.asciidoc

[[providers]]
== Providers

> Providers supply the key and value pairs that are used for variable substitution

Reviewer: You wouldn't have a way to know this, but we write key-value with a dash

```suggestion
Providers supply the key-value pairs that are used for variable substitution
```

and conditionals. Each provider's keys are automatically prefixed with the name
of the provider in the context of the {agent}.

For example, a provider named `foo` provides

> `{"key1": "value1", "key2": "value2"}`, the key and value pairs are placed in

Reviewer: ```suggestion
`{"key1": "value1", "key2": "value2"}`, the key-value pairs are placed in
```

[[context-providers]]
=== Context providers

Context providers give the current context of the running {agent}, for
example, agent information (ID, version), host information (hostname, IP
addresses), and environment information (environment variables).

> They can only provide a single key and value mapping. Think of them as singletons;

Reviewer: ```suggestion
They can only provide a single key-value mapping. Think of them as singletons;
```

> an update of a key and value mapping results in a re-evaluation of the entire

Reviewer: ```suggestion
an update of a key-value mapping results in a re-evaluation of the entire
```

> configuration. These providers are normally very static, but not

Author: This doesn't make sense to me

Reviewer: I think this is OK as-is.

required. A value can change which results in re-evaluation. 

> Context providers use ECS naming to ensure consistency and understanding throughout documentation and projects. 

Author: Should ECS be spelled out and then followed by (ECS)? I didn't see this in the style guide.

Reviewer: Yes it should. ECS stands for Elastic common schema. The rewrite makes sense to me.
```suggestion
Context providers use the Elastic Common Schema (ECS) naming to ensure consistency and understanding throughout documentation and projects. 
```

[[env-provider]]
==== Env Provider

> Provides access to the environment variables as key and values.

Reviewer: ```suggestion
Provides access to the environment variables as key-value pairs.
```

For example, set the variable `foo`:

[source,shell]
----
foo=bar elastic-agent run
----

The environment variable can be referenced as `${env.foo}`.

[discrete]
[[kubernetes_secrets-provider]]
==== Kubernetes Secrets Provider

Provides access to the Kubernetes Secrets API.

The provider needs a `kubeconfig` file to establish connection to the Kubernetes API.

> It can automatically reach the API if it's run in an inCluster environment (agent runs as Pod).

Author: Should pod be capitalized? And should agent be capitalized? The style guide didn't really give a clear direction for either of these, but I think this will need to be spot checked throughout based on your feedback.

Reviewer: Agent should be capitalized. Pod should not.

Reviewer: Agent is a product name, so we have an attribute that you can use. This way you don't even have to worry about it 😄. I've explained it in more detail in another comment.

```suggestion
It can automatically reach the API if it's run in an inCluster environment ({agent} runs as pod).
```

[source,yaml]
----
providers.kubernetes_secrets:
  #kube_config: /Users/elastic-agent/.kube/config
----

> Reference the Kubernetes Secrets variable as `${kubernetes_secrets.default.somesecret.value}`,

Author: Should secret be capitalized? I didn't think so but I wanted to ask before I changed it.

Reviewer: Looking at the Kubernetes documentation, yes, I think it should: https://kubernetes.io/docs/concepts/configuration/secret/

> Provides the option to enable leaderelection between a set of Agents

Author: Is leaderelection one word or two?

Reviewer: leaderelection looks to be one word to me.

In general that's correct. Elastic Agent is the product name in this instance, so we want to capitalize all instances of Agent.

Reviewer: We actually have an attribute for this. All instances of "Agent" can be replaced with `{agent}`. For the plural form, you can use `{agent}'s`. This will substitute our capitalized product name automatically.

running on Kubernetes. Only one Agent at a time will be the holder of the leader
lock and based on this, configurations can be enabled with the condition

> that the Agent holds the leadership. This is useful in cases where the agent between a set of agents collects cluster wide metrics for the Kubernetes cluster, such as the `kube-state-metrics` endpoint.

Author: I wanted you to approve this rephrase!

Reviewer: I think this is good

Provider needs a `kubeconfig` file to establish a connection to Kubernetes API.

> It can automatically reach the API if it's running in an inCluster environment (Agent runs as Pod).

Author: I wasn't sure if agent and pod should be capitalized here, similar as called out above.

Reviewer: See previous answer

[[dynamic-providers]]
=== Dynamic Providers

> Dynamic providers give an array of multiple key and value mappings. Each

Reviewer: ```suggestion
Dynamic providers give an array of multiple key-value mappings. Each
```

> key and value mapping is combined with the previous context provider's key and value

Reviewer: ```suggestion
key-value mapping is combined with the previous context provider's key and value
```

mapping which provides a new unique mapping that is used to generate a
configuration.

[discrete]
[[local-dynamic-provider]]
==== Local dynamic provider

> Define multiple keys and values to generate multiple configurations.

Reviewer: ```suggestion
Define multiple key-value pairs to generate multiple configurations.
```

|Log path of the container
|===

> For example, the Docker provider provides the following inventory:

Author: Does docker need to be capitalized?

Reviewer: Yes

|Image of the container
|===

> Fox example, the Kubernetes provider provides the following inventory:

Reviewer: This is a potential scenario, so I think we need to clarify

```suggestion
Fox example, if the Kubernetes provider provides the following inventory:
```

----

`node`:: (Optional) Specify the node to scope {agent} to in case it
cannot be accurately detected when running {agent} in host network
mode.
`cleanup_timeout`:: (Optional) Specify the time of inactivity before stopping the

> running configuration for a container. This is 60s by default.

Author: Does the 60s need to be in ticks?

Reviewer 2: Yes, I would place it in ticks.

`sync_period`:: (Optional) Specify the timeout for listing historical resources.
`kube_config`:: (Optional) Use the given config file as configuration for Kubernetes
client. If kube_config is not set, the KUBECONFIG environment variable will be
checked and will fall back to InCluster if not present.

> `scope`:: (Optional) Specify the level for autodiscover. `scope` can

Author: Does autodiscover need to be hyphenated?

Reviewer 2: Good question. No need to hyphenate.
